File: backend/src/services/skill_storage.py
"""In-memory skill package storage service."""

from typing import Optional

from ..models import SkillPackage


class SkillStorage:
    """In-memory storage for skill packages."""

    def __init__(self):
        self._packages: dict[str, SkillPackage] = {}

    def save(self, package: SkillPackage) -> None:
        """Save a skill package.

        Args:
            package: SkillPackage to save.
        """
        self._packages[package.id] = package

    def get(self, package_id: str) -> Optional[SkillPackage]:
        """Get a skill package by ID.

        Args:
            package_id: ID of the package to retrieve.

        Returns:
            SkillPackage if found, None otherwise.
        """
        return self._packages.get(package_id)

    def delete(self, package_id: str) -> bool:
        """Delete a skill package.

        Args:
            package_id: ID of the package to delete.

        Returns:
            True if deleted, False if not found.
        """
        if package_id in self._packages:
            del self._packages[package_id]
            return True
        return False

    def list_all(self) -> list[SkillPackage]:
        """List all skill packages.

        Returns:
            List of all stored skill packages.
        """
        return list(self._packages.values())

    def clear(self) -> None:
        """Clear all stored packages."""
        self._packages.clear()


# Global storage instance
_storage: Optional[SkillStorage] = None


def get_skill_storage() -> SkillStorage:
    """Get the global skill storage instance."""
    global _storage
    if _storage is None:
        _storage = SkillStorage()
    return _storage

File: packages/backend/src/services/skill.storage.ts
/**
 * In-memory skill package storage service.
 */

import type { SkillPackage } from '@skills-runtime/shared';

/**
 * Extended SkillPackage with binary content for in-memory storage.
 */
export interface StoredSkillPackage extends SkillPackage {
  /** Binary content of files for repacking */
  binaryContents?: Map<string, Uint8Array>;
  /** Modified text content of files */
  modifiedContents?: Map<string, string>;
}

/**
 * In-memory storage for skill packages.
 */
export class SkillStorage {
  private packages: Map<string, StoredSkillPackage> = new Map();

  /**
   * Save a skill package.
   *
   * @param pkg - SkillPackage to save.
   */
  save(pkg: StoredSkillPackage): void {
    this.packages.set(pkg.id, pkg);
  }

  /**
   * Get a skill package by ID.
   *
   * @param packageId - ID of the package to retrieve.
   * @returns SkillPackage if found, undefined otherwise.
   */
  get(packageId: string): StoredSkillPackage | undefined {
    return this.packages.get(packageId);
  }

  /**
   * Delete a skill package.
   *
   * @param packageId - ID of the package to delete.
   * @returns True if deleted, false if not found.
   */
  delete(packageId: string): boolean {
    return this.packages.delete(packageId);
  }

  /**
   * List all skill packages.
   *
   * @returns List of all stored skill packages.
   */
  listAll(): StoredSkillPackage[] {
    return Array.from(this.packages.values());
  }

  /**
   * Clear all stored packages.
   */
  clear(): void {
    this.packages.clear();
  }

  /**
   * Check if a package exists.
   *
   * @param packageId - ID of the package to check.
   * @returns True if exists, false otherwise.
   */
  has(packageId: string): boolean {
    return this.packages.has(packageId);
  }

  /**
   * Get count of stored packages.
   */
  get count(): number {
    return this.packages.size;
  }
}

// Global storage instance
let storage: SkillStorage | null = null;

/**
 * Get the global skill storage instance.
 */
export function getSkillStorage(): SkillStorage {
  if (!storage) {
    storage = new SkillStorage();
  }
  return storage;
}

/**
 * Reset the global skill storage (for testing).
 */
export function resetSkillStorage(): void {
  storage = null;
}
